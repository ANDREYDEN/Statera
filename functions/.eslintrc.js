module.exports = {
  root: true,
  env: {
    es6: true,
    node: true,
  },
  extends: [
    'eslint:recommended',
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',
    'google',
    'plugin:@typescript-eslint/recommended',
  ],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: ['tsconfig.json', 'tsconfig.dev.json'],
    sourceType: 'module',
  },
  ignorePatterns: [
    '/lib/**/*', // Ignore built files.
  ],
  plugins: ['@typescript-eslint', 'import'],
  rules: {
    'quotes': ['error', 'single'],
    'object-curly-spacing': ['warn', 'always'],
    'require-jsdoc': 'off',
    'semi': ['error', 'never'],
    'no-unused-vars': ['warn', { 'varsIgnorePattern': '^_', 'argsIgnorePattern': '^_' }],
    '@typescript-eslint/no-unused-vars': ['warn', { 'varsIgnorePattern': '^_', 'argsIgnorePattern': '^_' }],
    'max-len': ['warn', { 'code': 100 }],
    'import/no-unresolved': ['warn'],
    'indent': ['error', 2],
  },
}
