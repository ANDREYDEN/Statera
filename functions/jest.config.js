/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
module.exports = {
  rootDir: './',
  preset: 'ts-jest',
  testEnvironment: 'node',
  testMatch: [
    '**/tests/**/*.test.ts'
  ],
  maxWorkers: 1,
};