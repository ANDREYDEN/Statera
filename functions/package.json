{
  "name": "functions",
  "scripts": {
    "test": "jest",
    "lint": "eslint --ext .ts . --fix",
    "format": "prettier --write \"src/{**/*,*}.{js,ts,jsx,tsx,json}\"",
    "build": "tsc",
    "listen": "tsc --watch",
    "serve": "npm run build && firebase emulators:start --only functions",
    "shell": "npm run build && firebase functions:shell",
    "start": "npm run shell",
    "deploy": "firebase deploy --only functions",
    "logs": "firebase functions:log",
    "emulate": "export GOOGLE_APPLICATION_CREDENTIALS=\"./serviceAccount.json\" && firebase emulators:start",
    "listen-emulate": "npm run listen & npm run emulate"
  },
  "engines": {
    "node": "18"
  },
  "main": "lib/index.js",
  "dependencies": {
    "@google-cloud/vision": "^2.3.7",
    "@types/jest": "^29",
    "firebase-admin": "^12.1.1",
    "firebase-functions": "^5.0.1",
    "jest": "^29",
    "lodash": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "puppeteer": "^13.5.2",
    "ts-jest": "^29"
  },
  "devDependencies": {
    "@types/node-fetch": "^2.6.3",
    "@typescript-eslint/eslint-plugin": "^7",
    "@typescript-eslint/parser": "^7",
    "eslint": "^8",
    "eslint-config-google": "^0.14.0",
    "eslint-plugin-import": "^2.26.0",
    "firebase-functions-test": "^0.2.0",
    "prettier": "^3",
    "typescript": "^5.4.5"
  },
  "private": true,
  "optionalDependencies": {
    "fsevents": "^2.3.2"
  }
}