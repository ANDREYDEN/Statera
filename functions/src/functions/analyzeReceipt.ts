import * as vision from '@google-cloud/vision'
import { Product } from '../types/products'
import { defaultStore, stores } from '../types/stores'
import { verticalSegment } from '../utils'

export async function analyzeReceipt(
    receiptUrl: string,
    isWalmart: boolean, //TODO: deprecate
    storeName: string,
    withNameImprovement?: boolean
): Promise<Product[]> {
  console.log(`Analyzing receipt at ${receiptUrl}`)
  const client = new vision.ImageAnnotatorClient()

  console.log('Reading text from the image...')
  const [result] = await client.textDetection(receiptUrl)

  // first element contains information about all lines
  const labels = result.textAnnotations?.slice(1) ?? []
  console.log(labels.length > 0 ?
    `This image has some text: ${labels.length}` :
    'This image has no text')

  type LabelBox = { p1: number; p2: number; description: string }
  const lines: LabelBox[][] = []

  labels.forEach((label) => {
    const labelSegment = verticalSegment(label)
    const center = (labelSegment.p1 + labelSegment.p2) / 2
    const labelBox = { ...labelSegment, description: label.description ?? '' }

    for (const line of lines) {
      if (line[0].p1 < center && center < line[0].p2) {
        line.push(labelBox)
        return
      }
    }

    lines.push([labelBox])
  })

  const rows = lines.map((line) => line.map((label) => label.description))
  console.log('Raw image text data:', rows)

  const store = isWalmart ? stores.walmart : stores[storeName] ?? defaultStore

  let products = store.normalize(rows)
  products = store.filter(products)
  products = store.merge(products)

  if (withNameImprovement) {
    console.log('Applying name improvements...')
    products = await store.improveNaming(products)
  }

  return products
}
