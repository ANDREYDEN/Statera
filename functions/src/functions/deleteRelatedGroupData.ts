import * as admin from 'firebase-admin'

export async function deleteRelatedGroupData(groupId: string): Promise<void> {
  const expenseQuerySnap = await admin.firestore()
    .collection('expenses')
    .where('groupId', '==', groupId)
    .get()

  for (const expenseDoc of expenseQuerySnap.docs) {
    await expenseDoc.ref.delete()
  }
}
