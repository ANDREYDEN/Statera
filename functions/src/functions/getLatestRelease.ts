import * as admin from 'firebase-admin'
import fetch from 'node-fetch'

export async function getLatestRelease(platform: string): Promise<Release> {
    const listApps = platform == 'android'
        ? admin.projectManagement().listAndroidApps()
        : admin.projectManagement().listIosApps()
    const apps = await listApps
    if (!apps || apps.length === 0) throw new Error(`No ${platform} apps found`)

    const app = apps[0]
    const appId = app.appId
    const projectNumber = app.appId.split(':')[1]

    const accessToken = await admin.app().options.credential?.getAccessToken()
    const response = await fetch(
        `https://firebaseappdistribution.googleapis.com/v1/projects/${projectNumber}/apps/${appId}/releases`,
        {
            headers: {
                Authorization: `Bearer ${accessToken?.access_token}`,
            }
        })
    if (!response.ok) {
        const error = await response.text()
        throw new Error(`Error getting latest release: ${error}`)
    }
    const result = await response.json()
    const releases = result.releases
    if (!releases || releases.length === 0) throw new Error(`No ${platform} releases found`)

    return releases[0]
}

type Release = {
    displayVersion: string
}