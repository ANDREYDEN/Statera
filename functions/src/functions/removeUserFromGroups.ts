import * as admin from 'firebase-admin'
import { QueryDocumentSnapshot } from 'firebase-functions/v1/firestore'

export async function removeUserFromGroups(uid: string) {
  const app = admin.app()
  const groupsSnaps = await admin
      .firestore(app)
      .collection('groups')
      .where('memberIds', 'array-contains', uid)
      .get()
  for (const groupDoc of groupsSnaps.docs) {
    const memberIds = groupDoc.data()['memberIds']
    if (memberIds.length === 1) {
      await deleteGroup(groupDoc)
    } else {
      await removeUserFromGroup(uid, groupDoc)
    }
  }
}

async function removeUserFromGroup(uid: string, groupDoc: QueryDocumentSnapshot) {
  const app = admin.app()

  const memberIds = groupDoc.data()['memberIds']
  const members = groupDoc.data()['members']
  const balance = groupDoc.data()['balance']
  try {
    delete balance[uid]
    for (const id of Object.keys(balance)) {
      delete balance[id][uid]
    }
    const newMembers = members.filter((member: { uid: string }) => member.uid !== uid)
    const newMemberIds = memberIds.filter((id: string) => id !== uid)
    await admin
        .firestore(app)
        .collection('groups')
        .doc(groupDoc.id)
        .update({
          memberIds: newMemberIds,
          members: newMembers,
          balance,
        })
  } catch (e) {
    console.log(`Failed to remove user from group (${groupDoc.id}): ${e}`)
  }
}

async function deleteGroup(groupDoc: QueryDocumentSnapshot) {
  const app = admin.app()

  const expenses = await admin
      .firestore(app)
      .collection('expenses')
      .where('groupId', '==', groupDoc.id)
      .get()

  for (const expense of expenses.docs) {
    await admin.firestore(app).collection('expenses').doc(expense.id).delete()
  }
  await admin.firestore(app).collection('groups').doc(groupDoc.id).delete()
}
