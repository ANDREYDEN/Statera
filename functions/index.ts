import * as admin from 'firebase-admin'
import * as functions from 'firebase-functions'
import { firestoreBackup } from './src/admin'
import { analyzeReceipt } from './src/functions/analyzeReceipt'
import { removeUserFromGroups } from './src/functions/removeUserFromGroups'
import { updateUser } from './src/functions/updateUser'
import { UserData } from './src/types/userData'

admin.initializeApp()

export const scheduledBackup = firestoreBackup

export const setTimestampOnPaymentCreation = functions.firestore
    .document('payments/{paymentId}')
    .onCreate(async (snap, _) => {
      await snap.ref.update({ timeCreated: snap.createTime })
    })

export const getReceiptData = functions
    .runWith({
      timeoutSeconds: 300,
      memory: '4GB',
    })
    .https.onCall(async (data, _) => {
      if (!data.receiptUrl) {
        throw Error('The parameter receiptUrl is required.')
      }

      return analyzeReceipt(
          data.receiptUrl,
          data.isWalmart,
          data.storeName,
          data.withNameImprovement
      )
    })

export const cleanUpOnAccountDeletion = functions.auth
    .user()
    .onDelete(async (user, _) => {
      removeUserFromGroups(user.uid)
    })

export const changeUser = functions.firestore
    .document('users/{userId}')
    .onUpdate(async (change, context) => {
      const oldUserData = change.before.data() as UserData
      const newUserData = change.after.data() as UserData
      if (oldUserData.name !== newUserData.name 
        || oldUserData.photoURL !== newUserData.photoURL) {
        await updateUser(context.params.userId, newUserData)
      }
      return null
    })
