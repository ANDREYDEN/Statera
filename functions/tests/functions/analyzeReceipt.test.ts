import { analyzeReceipt } from '../../src/functions/analyzeReceipt'
import lcboReceiptData from '../__stubs__/lcbo_receipt_data.json'
import walmartReceiptData from '../__stubs__/walmart_receipt_data.json'
import walmartReceiptMediumData from '../__stubs__/walmart_receipt_medium/data.json'
import walmartReceiptMediumExpected from '../__stubs__/walmart_receipt_medium/expected.json'

const textDetection = jest.fn()
jest.mock('@google-cloud/vision', () => ({
  ImageAnnotatorClient: jest.fn().mockImplementation(function() {
    return { textDetection }
  }),
}))

describe('analyzeReceipt', () => {
  it('can analyze Walmart receipt', async () => {
    textDetection.mockResolvedValue(walmartReceiptData)
    const products = await analyzeReceipt('https://example.com', 'walmart')

    expect(products).toMatchSnapshot()
  })

  it('can analyze a medium Walmart receipt', async () => {
    textDetection.mockResolvedValue(walmartReceiptMediumData)
    const products = await analyzeReceipt('https://example.com', 'walmart')

    expect(products).toEqual(walmartReceiptMediumExpected)
  })

  it('can analyze LCBO receipt', async () => {
    textDetection.mockResolvedValue(lcboReceiptData)
    const products = await analyzeReceipt('https://example.com', 'lcbo')

    expect(products).toMatchSnapshot()
  })
})
