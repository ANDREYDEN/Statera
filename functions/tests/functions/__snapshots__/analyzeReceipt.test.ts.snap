// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyzeReceipt can analyze LCBO receipt 1`] = `
[
  {
    "deposit": 0.1,
    "id": "00491183",
    "name": "GEORGIAN BAY VODKA SMASH",
    "quantity": 1,
    "value": 3,
    "volume": 473,
  },
  {
    "deposit": 0.2,
    "id": "00028049",
    "name": "ZIRKOVA UNITY VODKA",
    "quantity": 1,
    "value": 34.75,
    "volume": 750,
  },
  {
    "deposit": 0.2,
    "id": "00446021",
    "name": "BACARDI COCONUT RUM",
    "quantity": 1,
    "value": 29,
    "volume": 750,
  },
  {
    "deposit": 0.1,
    "id": "00267286",
    "name": "SAUZA GOLD",
    "quantity": 1,
    "value": 20.85,
    "volume": 375,
  },
  {
    "deposit": 0.1,
    "id": "00019347",
    "name": "FOUNDER'S ORIGINAL BOURBON SOUR",
    "quantity": 1,
    "value": 2.85,
    "volume": 355,
  },
  {
    "deposit": 0.1,
    "id": "00024319",
    "name": "DILLON'S TANGERINE LEMON & A HINT OF M",
    "quantity": 1,
    "value": 3.05,
    "volume": 473,
  },
  {
    "deposit": 0.1,
    "id": "00024321",
    "name": "DILLON'S BLACKBERRY LEMON & A DASH OF",
    "quantity": 1,
    "value": 3.05,
    "volume": 473,
  },
  {
    "deposit": 0.1,
    "id": "00019350",
    "name": "FOUNDER'S ORIGINAL TEQUILA PALOMA",
    "quantity": 1,
    "value": 3,
    "volume": 355,
  },
]
`;

exports[`analyzeReceipt can analyze Walmart receipt 1`] = `
[
  {
    "name": "Yam / Sw Pot",
    "sku": "000000004074R",
    "value": 3.47,
  },
  {
    "name": "Lib Grk 650",
    "sku": "006568412394R",
    "value": 5.98,
  },
  {
    "name": "Hgdz Vncmlpr",
    "sku": "005500040742R",
    "value": 12,
  },
  {
    "name": "Chk Bst Bnls",
    "sku": "060538876501R",
    "value": 1.97,
  },
  {
    "name": "Cl Tuna Dill",
    "sku": "006136243483R",
    "value": 1.97,
  },
  {
    "name": "Cl Tuna - Chil",
    "sku": "006136243485R",
    "value": 1.97,
  },
  {
    "name": "Cl Tuna - Chil",
    "sku": "006136243485R",
    "value": 2.27,
  },
  {
    "name": "Mini Carrots",
    "sku": "003338366602R",
    "value": 1.47,
  },
  {
    "name": "Cucumber",
    "sku": "000000004593R",
    "value": 2.52,
  },
  {
    "name": "Sqursh Zuce",
    "sku": "0000000040678",
    "value": 2.52,
  },
  {
    "name": "Kiwi",
    "sku": "000000004030R",
    "value": 2.61,
  },
  {
    "name": "Wldrf Salad",
    "sku": "088804800082R",
    "value": 6.97,
  },
  {
    "name": "Onion Scal",
    "sku": "000000004068R",
    "value": 0.97,
  },
  {
    "name": "Broc Stlk",
    "sku": "000000004060R",
    "value": 2.47,
  },
  {
    "name": "Red Grapes",
    "sku": "000000004023R",
    "value": 3.75,
  },
]
`;
